src/examples/cuda_wrapper.c

```diff
@@ -0,0 +1,36 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <cuda_runtime.h>
+#include <HYPRE_config.h>
+
+typedef size_t devptr_t;
+
+#if defined(HYPRE_USING_CUDA)
+
+#define CUDA_MALLOC_MANAGED  device_malloc_managed_
+#define CUDA_FREE            device_free_
+
+int CUDA_MALLOC_MANAGED (const int *nbytes, devptr_t *devicePtr)
+{
+    void *tPtr;
+    int retVal;
+    retVal = (int) cudaMallocManaged (&tPtr, *nbytes, cudaMemAttachGlobal);
```

Comment:
Is this how Fortran users deal with allocating device memory, call a C interface?

Reply:
I guess yes, although I am not an expert. There is a more neat approach that doesn't need C wrapper such as
(https://docs.nvidia.com/hpc-sdk/compilers/cuda-fortran-prog-guide/index.html#cfexam-matrix-mult)
```
real, device, allocatable, dimension(:,:) :: Adev,Bdev,Cdev
```
My understanding is that this requires special Fortran compilers. Maybe we ask this to the users.

Reply:
I saw that, but I agree it doesn't look to be a portable solution.  I also agree that we should get some input here from our Fortran GPU users.

Reply:
I have used CUDA Fortran, and OpenACC combined with some features of CUDA Fortran for managing memory. This is supported by the `nvfortran` compiler from the [NVIDIA HPC SDK](https://developer.nvidia.com/hpc-sdk). I'm not sure if there is an easy, portable solution for deploying Fortran code on GPUs...

Indeed one can declare and allocate device/managed memory variables as follows:
```fortran
  real, device, allocatable, dimension(:,:) a_d
  (...)
  allocate(a_d)
```
replacing `device` with `managed` would declare it as managed memory.

alternatively, one can do as follows, which is at times convenient:
```fortran
  real, allocatable, dimension(:,:) a
#if defined(USE_CUDA)
  attributes(managed) :: a ! or device
#endif
  (...)
  allocate(a)
```
(a short CUDA Fortran reference [document](https://www.pgroup.com/lit/literature/pgi-cuf-pg-2018.pdf)).

Reply:
I'd be interested in getting an OpenMP4.5 version of one of these examples tested with this PR.  We use "!$omp map data..." pragmas to create and move data onto the GPU.  Its likely that users of this interface with have their own solution but so far the most portable solution for GPUs for fortran code has been OpenMP. 

```fortran
! Declare and allocate on the CPU
real, allocatable, dimension(:,:) a
allocate(a(n,m))

! Allocate device memory
#if defined(USE_GPU)
!$omp target enter data map(alloc:a)
#endif

! Use device here
...
!$omp target data use_device_ptr(a)
CALL call_hypre_routine( a ) 
!$omp end target data
...

! Clean-up device memory
#ifdefined(USE_GPU)
!$omp target exit data map(release:a)
#endif

```

Reply:
Thanks @p-costa and @flow-phys for your advice and input on this!